kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/listener/CollectingTestEngineListener.kt

```diff
@@ -16,31 +17,38 @@ class CollectingTestEngineListener : AbstractTestEngineListener(), Mutex by Mute
    val names = mutableListOf<String>()
    var errors = false
 
+   /**
+    * An active [Job] that will be completed when [engineFinished] is invoked.
+    * @see waitForEngineFinished
+    */
+   private val engineFinishedJob = Job()
```

Comment:
This is an interesting solution, using a Job as a boolean-like state which one can wait for while suspending.

It took me some time to decipher, as I had to make sure that the Job
* does not perform actual work, and
* does not have children (in which case `complete()` would not make `join()` return immediately).

Could you add some comment to that effect?

([The same could be achieved with a `Mutex` (playground)](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMi4wLjAiLCJwbGF0Zm9ybSI6ImphdmEiLCJhcmdzIjoiIiwibm9uZU1hcmtlcnMiOnRydWUsInRoZW1lIjoiaWRlYSIsImNvZGUiOiJpbXBvcnQga290bGlueC5jb3JvdXRpbmVzLmRlbGF5XG5pbXBvcnQga290bGlueC5jb3JvdXRpbmVzLmxhdW5jaFxuaW1wb3J0IGtvdGxpbnguY29yb3V0aW5lcy5ydW5CbG9ja2luZ1xuaW1wb3J0IGtvdGxpbnguY29yb3V0aW5lcy5zeW5jLk11dGV4XG5pbXBvcnQga290bGlueC5jb3JvdXRpbmVzLnN5bmMud2l0aExvY2tcbmltcG9ydCBrb3RsaW4udGltZS5EdXJhdGlvbi5Db21wYW5pb24uc2Vjb25kc1xuXG5mdW4gbWFpbigpOiBVbml0ID0gcnVuQmxvY2tpbmcge1xuICAgIHZhbCBlbmdpbmVGaW5pc2hlZCA9IE11dGV4KGxvY2tlZCA9IHRydWUpXG5cbiAgICBsYXVuY2gge1xuICAgICAgICBkZWxheSgyLnNlY29uZHMpXG4gICAgICAgIGVuZ2luZUZpbmlzaGVkLnVubG9jaygpXG4gICAgICAgIHByaW50bG4oXCJmaW5pc2hlZFwiKVxuICAgIH1cblxuICAgIGxhdW5jaCB7XG4gICAgICAgIGRlbGF5KDAuNS5zZWNvbmRzKVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMTogd2FpdGluZ1wiKVxuICAgICAgICBlbmdpbmVGaW5pc2hlZC53aXRoTG9jayB7fVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMTogZGV0ZWN0ZWQgZmluaXNoXCIpXG4gICAgfVxuXG4gICAgbGF1bmNoIHtcbiAgICAgICAgZGVsYXkoMS5zZWNvbmRzKVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMjogd2FpdGluZ1wiKVxuICAgICAgICBlbmdpbmVGaW5pc2hlZC53aXRoTG9jayB7fVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMjogZGV0ZWN0ZWQgZmluaXNoXCIpXG4gICAgfVxuXG4gICAgbGF1bmNoIHtcbiAgICAgICAgZGVsYXkoMy5zZWNvbmRzKVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMzogd2FpdGluZ1wiKVxuICAgICAgICBlbmdpbmVGaW5pc2hlZC53aXRoTG9jayB7fVxuICAgICAgICBwcmludGxuKFwib2JzZXJ2ZXIgMzogZGV0ZWN0ZWQgZmluaXNoXCIpXG4gICAgfVxufVxuIn0=), but that would also require some comment explaining to its uncommon use. I have a similar thing called [`Gate` in this KotlinConf talk companion repo](https://github.com/OliverO2/kmp-e2e-stress-testing-example/blob/main/application/library-testing/src/commonMain/kotlin/library/testing/Gate.kt).)

Reply:
Ahh yes, the `mutex.withLock {}` is probably better. I'll update the code and with a comment.